enhancements/single-node-production-edge-cluster-profile.md

```diff
@@ -0,0 +1,380 @@
+---
+title: single-node-production-edge-cluster-profile
+authors:
+  - "@dhellmann"
+  - "@eranco"
+  - "@romfreiman"
+  - "@markmc"
+reviewers:
+  - TBD, probably all leads
+approvers:
+  - "@derekwaynecarr"
+  - "@smarterclayton"
+creation-date: 2020-10-15
+last-updated: 2020-10-15
+status: implementable
+see-also:
+  - "/enhancements/update/cluster-profiles.md"
+  - "/enhancements/single-node-developer-cluster-profile.md"
+  - https://github.com/openshift/enhancements/pull/302
+  - https://github.com/openshift/enhancements/pull/414
+  - https://github.com/openshift/enhancements/pull/440
+---
+
+# Single-node Production Edge Cluster Profile
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Add a new `single-node-production-edge` [cluster
+profile](https://github.com/openshift/enhancements/blob/master/enhancements/update/cluster-profiles.md)
+for *production use* in "edge" deployments on servers that are not
+considered to be resource-constrained.
+
+A cluster deployed using this profile will differ from the default
+`self-managed-highly-available` cluster profile in several significant
+ways:
+
+* The single node serves as both the cluster’s control plane and as a
+  worker node.
+* Many operators will be configured to reduce the footprint of their
+  operands, such as by running fewer replicas.
+* In-place upgrades will not be supported by the first iteration of
```

Comment:
Yeah, that's the general idea. These sorts of deployments are expected to be part of a larger system that includes an orchestration tool outside of the cluster for managing 100s or 1000s of individual instances. Assuming a wipe-and-rebuild approach can be implemented within the other constraints like the length of change windows, it seemed reasonable to go that route instead of trying to make in-place work. On the other hand, if in-place isn't a big deal, maybe we don't need to make that assumption. So, it's definitely still up for discussion.

Reply:
just noting there is a consideration for remote low bandwidth clusters in which we'd want to update only the delta vs. the entire image, which may be more feasible for an in-place upgrade vs. a full re-image.

Reply:
I think the low bandwidth is not part of the enhancement. Here we should agree how single node looks like. We will deal with farther improvements (bandwidth, bootstrap removal) as part of other enhancements.

Reply:
Regrading the delta - I assume that we dont recreate all the containers images every release - so it should be solved by docker. Unless I'm wrong.
Regarding rchos - I have no clue - is there such functionality in place? @cgwalters ?

Reply:
One reason ostree is popular is quite a while ago we implemented a pretty good delta mechanism.  We aren't using it for RHCOS but "base" FCOS does use it.

https://ostreedev.github.io/ostree/formats/#static-deltas

There'd be some work to enable this but not really hard.

One even more radical approach here would be to commit all of the container images into ostree as well - lifecycle bind everything into a single transactional update.   Basically put the containers in e.g. `/usr/share/containers/<sha256>` and teach crio how to use the equivalent of `podman --root=/usr/share/containers/<sha256>` if it exists.  Then we get ostree deltas for everything. 

(But the "lifecycle binding" is also important - this way we have the old OS with old containers, or new OS with new containers, and no possibility of skew)

This is what some ostree users are doing today, although at least one I talked to was using `systemd-nspawn` - but same principle.

There's also https://blogs.gnome.org/alexl/2020/05/13/putting-container-updates-on-a-diet/ which is a clearer long term across-the-board win; not sure what the status of it is though.